roles/ora-host/tasks/main.yml

```diff
@@ -67,7 +67,6 @@
     name: "{{ oracle_required_rpms }}"
     state: present
     lock_timeout: 180
-    enablerepo: rhel-7-server-optional-rpms
```

Comment:
This repo is definitely required with RHEL7;  if testing with another OS, can you make a change that continues to work with RHEL7?

Reply:
I think the problem is with rhel-7-server-optional-rpms. It's not defined in any repo file under /etc/yum.repos.d in BMS servers or RHEL VMs in GCP. When the task tries to enable this repo it fails.

Reply:
Manually tried to trigger this on a vanilla server. Details follow:

I am going through an install in a newly reimaged RHEL7.7 server and was able to see that for this RHEL server registered with RedHat, refreshing the redhat.repo (empty redhat.repo and yum repolist) will result in:
```
   2213 [rhel-7-server-optional-rpms]
   2214 metadata_expire = 86400
   2215 enabled_metadata = 0
   2216 sslclientcert = /etc/pki/entitlement/2455310734525423220.pem
   2217 baseurl = https://cdn.redhat.com/content/dist/rhel/server/7/$releasever/$basearch/optional/os
   2218 ui_repoid_vars = releasever basearch
   2219 sslverify = 1
   2220 name = Red Hat Enterprise Linux 7 Server - Optional (RPMs)
   2221 sslclientkey = /etc/pki/entitlement/2455310734525423220-key.pem
   2222 gpgkey = file:///etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release
   2223 enabled = 0
   2224 sslcacert = /etc/rhsm/ca/redhat-uep.pem
   2225 gpgcheck = 1
```

When that repo was enabled,

```
[root@at-00010-svr002 yum.repos.d]# vi redhat.repo
[root@at-00010-svr002 yum.repos.d]# yum repolist
Loaded plugins: product-id, search-disabled-repos, subscription-manager
...
rhel-7-server-optional-rpms/7Server/x86_64                                                                                                                                                  | 3.2 kB  00:00:00     
rhel-7-server-optional-rpms/7Server/x86_64/group                                                                                                                                            |  22 kB  00:00:00     
rhel-7-server-optional-rpms/7Server/x86_64/updateinfo                                                                                                                                       | 2.9 MB  00:00:01     
rhel-7-server-optional-rpms/7Server/x86_64/primary_db                                                                                                                                       | 9.8 MB  00:00:03     
rhel-7-server-rpms/7Server/x86_64                                                                                                                                                           | 3.5 kB  00:00:00     
repo id                                                                                           repo name                                                                                                  status
epel/7Server/x86_64                                                                               EPEL YUM repo                                                                                                   0
rhel-7-server-optional-rpms/7Server/x86_64                                                        Red Hat Enterprise Linux 7 Server - Optional (RPMs)                                                        22,814
rhel-7-server-rpms/7Server/x86_64                                                                 Red Hat Enterprise Linux 7 Server (RPMs)                                                                   31,594
repolist: 54,408
[root@at-00010-svr002 yum.repos.d]# 

```

Reply:
Yes probably for Vanilla RHEL server it would work but for any BMS server (RHEL or OEL) it always fails at this point. I have just received this error from a customer doing a single instance installation on  RHEL BMS server.

Reply:
Thank you, can you please provide details on the error received?

The reimaged server I am installing is indeed a BMS host located in Sydney.

Reply:
My pleasure :)

This is the error both me and the customer received unless enabling both repos is hashed

enablerepo: rhel-7-server-optional-rpms
enablerepo: rhui-rhel-7-server-rhui-optional-rpms
 
2020-11-10 17:30:51,616 p=13691 u=ansible n=ansible | TASK [ora-host : Install Oracle required packages (base/non-rhui config) Redhat] *****************************************************
2020-11-10 17:30:52,462 p=13691 u=ansible n=ansible | fatal: [00001-svr001]: FAILED! => {"changed": false, "msg": "Repository rhel-7-server-optional-rpms not found."}
2020-11-10 17:30:52,567 p=13691 u=ansible n=ansible | fatal: [00001-svr002]: FAILED! => {"changed": false, "msg": "Repository rhel-7-server-optional-rpms not found."}

Reply:
Thanks and the OS version/kernel details please...

Reply:
OEL 7.9 4.14.35-2025.401.4.el7uek.x86_64
RHEL 7.9 (I don't have the customer's kernel release)

Reply:
I got a freshly reimaged OEL 7.9 BMS machine and I don't see the repo files having the *optional* repos.
 ```
[root@OEL7.9server yum.repos.d]# ll
total 12
-rw-r--r--. 1 root root 4075 Mar  3 03:28 oracle-linux-ol7.repo
-rw-r--r--. 1 root root 2587 Mar  3 03:27 uek-ol7.repo
-rw-r--r--. 1 root root  226 Mar  3 03:27 virt-ol7.repo
[root@OEL7.9server yum.repos.d]# yum repolist
Loaded plugins: ulninfo
https://yum.oracle.com/repo/OracleLinux/OL7/UEKR6/x86_64/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: yum.oracle.com; Unknown error"
Trying other mirror.
https://yum.oracle.com/repo/OracleLinux/OL7/UEKR6/x86_64/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: yum.oracle.com; Unknown error"
Trying other mirror.
https://yum.oracle.com/repo/OracleLinux/OL7/latest/x86_64/repodata/repomd.xml: [Errno 14] curl#6 - "Could not resolve host: yum.oracle.com; Unknown error"
Trying other mirror.
repo id                                                                repo name                                                                                                                             status
ol7_UEKR6/x86_64                                                       Latest Unbreakable Enterprise Kernel Release 6 for Oracle Linux 7Server (x86_64)                                                      0
ol7_latest/x86_64                                                      Oracle Linux 7Server Latest (x86_64)                                                                                                  0
repolist: 0
[root@OEL7.9server yum.repos.d]# cat oracle-linux-ol7.repo 
[ol7_latest]
name=Oracle Linux $releasever Latest ($basearch)
baseurl=https://yum$ociregion.$ocidomain/repo/OracleLinux/OL7/latest/$basearch/
gpgkey=file:///etc/pki/rpm-gpg/RPM-GPG-KEY-oracle
gpgcheck=1
enabled=1
...
[root@OEL7.9server yum.repos.d]# grep -i optional *
oracle-linux-ol7.repo:[ol7_optional_latest]
oracle-linux-ol7.repo:name=Oracle Linux $releasever Optional Latest ($basearch)
oracle-linux-ol7.repo:baseurl=https://yum$ociregion.$ocidomain/repo/OracleLinux/OL7/optional/latest/$basearch/
oracle-linux-ol7.repo:[ol7_optional_archive]
oracle-linux-ol7.repo:name=Oracle Linux $releasever Optional ($basearch) - Archive
oracle-linux-ol7.repo:baseurl=https://yum$ociregion.$ocidomain/repo/OracleLinux/OL7/optional/archive/$basearch/
```

Also, verified that the `when:` will be triggered because:
```
[jcnarasimhan@control-node ~]$ ansible all -i <IP of OEL7.9 host>, -m setup -u customeradmin --ask-pass > 004_ansible_facts.txt
SSH password: 

[jcnarasimhan@control-node ~]$ grep -i os_family 004_ansible_facts.txt 
        "ansible_os_family": "RedHat", 
```

Will come up with a fix/logic that bypasses this.

Reply:
Was able to reproduce this on RHEL7.7. Details follow.

Error:
```
TASK [ora-host : Install Oracle required packages (base/non-rhui config)] ******
fatal: [at-00010-svr002]: FAILED! => {"changed": false, "msg": "Repository rhel-7-server-optional-rpms not found."}
```

The issue in my case was that I changed the time on the BMS host to a different tz, that messed up the `redhat.repo`'s subscription. 

Fix:
```
[root@bms_server_2 yum.repos.d]# subscription-manager unregister
Unregistering from: subscription.rhsm.redhat.com:443/subscription
System has been unregistered.

[root@bms_server_2 yum.repos.d]# subscription-manager register
Registering to: subscription.rhsm.redhat.com:443/subscription
Username: google-bms
Password: 
The system has been registered with ID: 5f113c06-d888-4f5e-bdc6-0318531f5e08
The registered system name is: bms_server_2

[root@bms_server_2 yum.repos.d]# 
[root@bms_server_2 yum.repos.d]# yum repolist
Loaded plugins: product-id, search-disabled-repos, subscription-manager
This system is registered with an entitlement server, but is not receiving updates. You can use subscription-manager to assign subscriptions.
repolist: 0

[root@bms_server_2 yum.repos.d]# subscription-manager attach --auto
Installed Product Current Status:
Product Name: Red Hat Enterprise Linux Server
Status:       Subscribed

[root@bms_server_2 yum.repos.d]# ll
total 204
-rw-r--r--. 1 root root   149 Apr 16 11:47 epellll.repo
-rw-r--r--. 1 root root   124 Apr 16 21:26 epel.repo
-rw-r--r--. 1 root root   316 Apr 16 21:13 epel.repo.worked
-rw-r--r--. 1 root root 97509 Apr 26 19:28 redhat.repo
-rw-r--r--. 1 root root 95868 Apr 16 19:23 redhat.repo.virgin

[root@bms_server_2 yum.repos.d]# yum repolist
Loaded plugins: product-id, search-disabled-repos, subscription-manager
repo id                                                                           repo name                                                                               status
!rhel-7-server-rpms/7Server/x86_64                                                Red Hat Enterprise Linux 7 Server (RPMs)                                                31,606
repolist: 31,606
```

Or, running the following ansible snippet from Marc's experimental resolves the issue:
```
[jcnarasimhan@control-node host-ansible]$ cat redhat-license.yml 
- hosts: all
  gather_facts: false
  tasks:
  - name: Remove registrations
    become: true
    redhat_subscription:
      state: absent
  - name: Register host
    become: true
    redhat_subscription:
      state: present
      username: google-bms
      password: "somepwd"
      auto_attach: true
```

Still, pending: coming up with a logic that bypasses this check on OEL 7.*.

Reply:
Brilliant, thank you :)

Reply:
Hello Ahmad,
Thanks for the finding on the line: `enablerepo: rhel-7-server-optional-rpms` that triggered a research comparing 2 OS platforms - RHEL 7.7 and OEL 7.9.

Please find below details of the logic arrived at for disabling the `rhel-7-server-optional-rpms` repo:
* We have a total of 52 rpms stored in the variable: `{{ oracle_required_rpms }}`.
   *  RHEL: Out of that 52, only one rpm `compat-libstdc++-33 ` comes from the repo `rhel-7-server-optional-rpms` in RHEL.
   * OEL: Out of that 52, all of them come from the repo `ol7_latest` in OEL.

* How many of the 52 are already installed?*
    * 22 in OEL
          * I didn't capture that in a RHEL BMS host before running our bms-toolkit there, but in OEL host, we have 22 already installed and 30 to be installed by our `ora-host` role.
* Logic summary based on the findings above: We need a fact that doesn’t break RHEL and at the same time works in OEL
    * For RHEL: We need `enablerepo: rhel-7-server-optional-rpms` only for RHEL.
    * For OEL: We still need to install all of the 52 rpms in `{{ oracle_required_rpms }}`, but we should not have to enable any repo.
* Facts available for OEL & RHEL:
   * OEL:
```
            "ansible_diff_mode": false,
            "ansible_distribution": "OracleLinux",
            "ansible_distribution_file_parsed": true,
            "ansible_distribution_file_path": "/etc/oracle-release",
            "ansible_distribution_file_search_string": "Oracle Linux",
            "ansible_distribution_file_variety": "OracleLinux",
            "ansible_distribution_major_version": "7",
            "ansible_distribution_release": "n/a",
            "ansible_distribution_version": "7.9",
```
   * RHEL:
```
            "ansible_distribution": "RedHat",
            "ansible_distribution_file_parsed": true,
            "ansible_distribution_file_path": "/etc/redhat-release",
            "ansible_distribution_file_search_string": "Red Hat",
            "ansible_distribution_file_variety": "RedHat",
            "ansible_distribution_major_version": "7",
            "ansible_distribution_release": "Maipo",
            "ansible_distribution_version": "7.7",
```
* Above logic translates to a code change something like the following:
```
- name: Install Oracle required packages (base/non-rhui config) for OEL
 yum:
   name: "{{ oracle_required_rpms }}"
   state: present
   lock_timeout: 180
   ##### enablerepo: rhel-7-server-optional-rpms
 when:
   - install_os_packages|bool and ansible_os_family == 'RedHat' and ansible_distribution == 'OracleLinux'
   - redhat_repo.stat.exists|bool
 tags: os-packages
 
- name: Install Oracle required packages (base/non-rhui config) for RHEL
 yum:
   name: "{{ oracle_required_rpms }}"
   state: present
   lock_timeout: 180
   enablerepo: rhel-7-server-optional-rpms
 when:
   - install_os_packages|bool and ansible_os_family == 'RedHat' and ansible_distribution == 'RedHat'
   - redhat_repo.stat.exists|bool
 tags: os-packages

```

Request you to please see if above makes sense and if it does, request you to incorporate in your PR please.

Thanks

Reply:
Hi JC, 

Thank you for the detailed analysis, I already tried this logic when I first hit the error in OEL (to only enable the repo when the ansible_distribution is Redhat). To my surprise enabling the repo gives the same error even in RHEL.

Reply:
Hello Ahmad,

If it's possible to reproduce the above in RHEL in your environment, we can collaborate via gvc to shoot for the fix in RHEL, but in my case I wasn't able to reproduce in a newly reimaged 7.7 RHEL.

But in OEL7.9, I was able to simulate missing rpms and fix it.
There was a logical error in the above snippet in my prior update in the `when` clause for OEL.

The correct block that was tested successfully is as follows:
```
- name: Install Oracle required packages (base/non-rhui config) for RHEL
  yum:
    name: "{{ oracle_required_rpms }}"
    state: present
    lock_timeout: 180
    enablerepo: rhel-7-server-optional-rpms
  when:
    - install_os_packages|bool and ansible_os_family == 'RedHat' and ansible_distribution == 'RedHat'
    - redhat_repo.stat.exists|bool
  tags: os-packages

- name: Stat OEL7 repo
  stat:
    path: /etc/yum.repos.d/oracle-linux-ol7.repo
  register: oraclelinux_repo

- name: Install Oracle required packages (base/non-rhui config) for OEL
  yum:
    name: "{{ oracle_required_rpms }}"
    state: present
    lock_timeout: 180
  when:
    - install_os_packages|bool and ansible_os_family == 'RedHat' and ansible_distribution == 'OracleLinux'
    - oraclelinux_repo.stat.exists|bool
  tags: os-packages

```

Based on the above and notes shared via (internal): https://docs.google.com/document/d/1mX59YHO37MavCGmfLlTtEmitRc53oeYkEgU3PtYMbrg/edit?usp=sharing ,

I will create a separate PR just with above code fix that will be fixing missing rpms in OEL7.9.
Please let me know if you have any inputs on reproducing it in RHEL and we can collaborate together.

Thanks

Reply:
Closing the loop on this here that the following 2 PRs address the points identified in PRs #59 and #60. 

- firewall settings needed for inter-node communication #75
- rpm related config changes tested successfully for OEL 7.9 #76